src/flow_graph.py

```diff
@@ -1013,10 +1032,169 @@ def terminate_infinite_loops(nodes: List[Node]) -> None:
+
+    def __len__(self) -> int:
+        return len(self.refs)
+
+
+@dataclass
+class AccessRefs:
+    refs: Dict[Access, RefSet] = field(default_factory=dict)
+
+    @staticmethod
+    def initial_function_regs(args: List[Access], arch: ArchFlowGraph) -> "AccessRefs":
+        info = AccessRefs()
+        # Set all the registers that are valid to access at the start of a function
+        for r in arch.saved_regs:
+            info.refs[r] = RefSet.special(f"saved_{r}")
+        for r in arch.constant_regs:
+            info.refs[r] = RefSet.special(f"const_{r}")
+        for a in args:
+            info.refs[a] = RefSet.special(f"arg_{a}")
+
+        info.refs[arch.return_address_reg] = RefSet.special(f"return")
+        info.refs[arch.stack_pointer_reg] = RefSet.special(f"sp")
+
+        return info
+
+    def get(self, reg: Access) -> RefSet:
+        return self.refs.get(reg, RefSet.invalid())
+
+    def add(self, reg: Access, ref: Reference) -> None:
+        if reg not in self:
+            self.refs[reg] = RefSet([ref])
+        else:
+            self.refs[reg].add(ref)
+
+    def extend(self, reg: Access, refs: RefSet) -> None:
+        if reg not in self:
+            self.refs[reg] = refs.copy()
+        else:
+            self.refs[reg].update(refs)
+
+    def remove(self, reg: Access) -> None:
+        self.refs.pop(reg)
+
+    def remove_ref(self, ref: Reference) -> None:
+        to_remove = []
+        for access, reflist in self.refs.items():
+            if ref in reflist:
+                reflist.remove(ref)
+            if not reflist:
+                to_remove.append(access)
+        for access in to_remove:
+            self.refs.pop(access)
+
+    def is_empty(self) -> bool:
+        return not self.refs
+
+    def copy(self) -> "AccessRefs":
+        return AccessRefs(refs=self.refs.copy())
+
+    def update(self, other: "AccessRefs") -> None:
+        self.refs.update(other.refs)
+
+    def values(self) -> ValuesView[RefSet]:
+        return self.refs.values()
+
+    def items(self) -> ItemsView[Access, RefSet]:
+        return self.refs.items()
+
+    def __contains__(self, key: Access) -> bool:
+        return key in self.refs
+
+    def __iter__(self) -> Iterator[Access]:
+        return iter(self.refs)
+
+    def __setitem__(self, key: Access, value: RefSet) -> None:
+        self.refs[key] = value
+
+
 @dataclass(frozen=True)
 class FlowGraph:
     nodes: List[Node]
 
+    # The set of all phi Accesses needed for each Node
+    node_phis: Dict[Node, AccessRefs] = field(default_factory=dict)
+    # The set of source References for each input Access to each Instruction
+    instr_inputs: Dict[InstrRef, AccessRefs] = field(default_factory=dict)
+    # The set of all downstream Instructions whose inputs depend on outputs from a given
+    # Instruction. This is the same information as `instr_inputs`, but reversed.
+    instr_references: Dict[InstrRef, AccessRefs] = field(default_factory=dict)
```

Comment:
I imagine you considered making these properties of Node and Instruction respectively, but it didn't work out?

Reply:
I moved `node_phis` into `Node`, but I don't know about moving the other fields into `Instruction`.
(Right now `InstrRef` references `Reference` which references `Node`, so at a minimum, there's that dependency chain to flatten)